Add Azure Workload Identity Auth Provider

akashsinghal opened the pull request:
- Adds Workload Identity Support for AKS and independent clusters
- Refer to design [document](https://hackmd.io/hs8IFrw5Sx66gOWlrICsig?both) for guide on how to setup ACR-enabled AKS Workload Identity

Note: Requires #123 to be merged

Closes #120

susanshi commented on pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -0,0 +1,168 @@
+	artifactHostName := parsedSpec.Hostname()
+
+	// retrieve an AAD Access token from ARM
+	aadToken, err := getAADAccessToken(tenantID, "https://management.azure.com/")
```
ACR is moving away from arm audienced AAD token, can can pass the ACR audience instead https://containerregistry.azure.net

sajayantony commented:
/cc @mnltejaswini

susanshi reviewed (commented):
Thanks Akash, looks good overall, just left some questions.

mnltejaswini commented on pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -0,0 +1,158 @@
+)
+
+type AzureWIProviderFactory struct{}
+type azureWIAuthProvider struct{}
```
Please refer to https://go.dev/doc/effective_go#names (a supplementary reference https://pthethanh.herokuapp.com/blog/articles/golang-name-conventions)  for naming conventions

mnltejaswini commented on pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -0,0 +1,158 @@
+// Uses Azure Workload Identity to retrieve an AAD access token which can be
+// exchanged for a valid ACR refresh token for login.
+func (d *azureWIAuthProvider) Provide(artifact string) (provider.AuthConfig, error) {
+	tenantID := os.Getenv("AZURE_TENANT_ID")
```
Do you need to validate if the value is not empty?

akashsinghal replied:
Yes I'll add this validation to the enabled method

akashsinghal commented on pkg/referrerstore/oras/authprovider/azure/azure_helper.go:
```diff
@@ -0,0 +1,266 @@
+
+const (
+	userAgentHeader = "User-Agent"
+	userAgent       = "kubernetes-credentialprovider-acr"
```
The azure helper methods are directly taken from the kubernetes-credentialprovider implementation published by the out-of-tree azure provider. I think it's fine to set it to be ratify specific but I'll try it out.

mnltejaswini commented on pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	// .default needs to be added to the scope
+	if !strings.HasSuffix(resource, ".default") {
```
Isn't this resource static in our case? Do you need this to build for every call?

akashsinghal replied:
Yes you're right. In fact, what we can actually do is frontload the getAADToken operation to happen upon auth provider registration on pod startup since the AADtoken is not registry host name specific in this scenario.

mnltejaswini commented on pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -0,0 +1,158 @@
+		resource += "/.default"
+	}
+
+	result, err := confidentialClientApp.AcquireTokenByCredential(context.Background(), []string{resource})
```
Isn't there any parent context that can be passed here rather than constructing new context?

mnltejaswini commented on pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -0,0 +1,158 @@
+	authorityHost := os.Getenv("AZURE_AUTHORITY_HOST")
+
+	// read the service account token from the filesystem
+	signedAssertion, err := readJWTFromFS(tokenFilePath)
```
How often these envs change? For example how often the token file changes? If they are injected when the pod is provisioned, do you need to read this file for every request?

akashsinghal replied:
I believe the token file is injected into the pod on provision. I'll need to look into it to confirm. If that's the case, then we can handle the AAD token operation on pod startup in auth provider registration process rather than on each Provide call.

mnltejaswini replied:
Is this still pending?

akashsinghal replied:
I have implemented the change. Now the initial AAD token exchange happens on pod startup and not every time Provide is called. I also added an ExpiresOn field for the credentials returned to account for the 24 hour AAD token life span. I also added logic to check for token expiry and call a new AAD token exchange as necessary.

akashsinghal commented on pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -0,0 +1,158 @@
+
+// Create returns an AzureWIAuthProvider
+func (s *AzureWIProviderFactory) Create(authProviderConfig provider.AuthProviderConfig) (provider.AuthProvider, error) {
+	conf := azureWIAuthProviderConf{}
```
conf is used when unmarshalling the auth provider config json. Line 56.

akashsinghal replied:
The name field must exist so I'll validation for this

mnltejaswini commented on pkg/referrerstore/oras/authprovider/azure/azure_helper.go:
```diff
@@ -0,0 +1,266 @@
+	var err error
+	data := url.Values{
+		"service":       []string{directive.service},
+		"grant_type":    []string{"access_token_refresh_token"},
```
Why is the access token passed as refresh token? can the grant type be access token?

mnltejaswini commented on pkg/referrerstore/oras/authprovider/azure/azure_helper.go:
```diff
@@ -0,0 +1,266 @@
+	}
+
+	var realmURL *url.URL
+	if realmURL, err = url.Parse(directive.realm); err != nil {
```
Do you want to do this first before creating the data

akashsinghal commented on pkg/referrerstore/oras/authprovider/azure/azure_helper.go:
```diff
@@ -0,0 +1,266 @@
+}
+
+// formats and sends a challenge request to the server address
+func receiveChallengeFromLoginServer(serverAddress, scheme string) (*authDirective, error) {
```
oh wow that so much simpler. I'll switch to using the Azure SDK for this.

akashsinghal commented:
@susanshi @mnltejaswini Thanks for your reviews! I have gone through and made extensive changes to address the feedback. Here's a summary of the updates:

1. Removed existing registry auth exchange functionality. Switched to use Azure-sdk-for-go's container registry client for auth token exchange.
2. Moved MSAL token file-AAD access token exchange to happen on pod startup rather than for every Provide call. The AAD scoped token can be acquired before since it's not dependent on the specific registry. (This improves the timeout failures. However, it still sometimes fails to complete within the 3 seconds, but at least now it doesn't fail all the time)
3. Added ExpiresOn field to AuthConfig which can be used to set the lifetime of the creds. This is primarily needed to handle AAD token's 24hr lifespan. The current cached registry client map needs to be refreshed once the AAD token expires. 
4. Added context field to Enabled and Provide methods in the interface. Allows server's context to now be used throughout auth providing process.
5. Added some basic sanity tests for workload identity provider.

susanshi reviewed (approved):
thanks @akashsinghal , looks good to me. Left some minor suggestions

sajayantony commented:
Recommend merging and open any feedback as follow up issues or PR.
